### PR TITLE
[connector/servicegraph] Fix failed label does not work leads to servicegraph metrics error

### LINKED ISSUE #32018: [connector/servicegraph] failed label does not work leads to servicegraph metrics error
### Component(s)

connector/servicegraph

## Description
The `failed` label fails to distinguish the succeed and failed edge in servicegraph metrics. 

I found this weird graph metrics:
image

The two metrics have the same label set except `failed`, their values are very close(during six hours) which is impossible. 

Besides, the `traces_service_graph_request_total` contains a label `failed=true` which also looks like a bug.

After reading the component code, I found the `failed` dimension doesn't join into the `metricKey`:
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/0a12edebb5aad2657c7401c50136d54acc89e573/connector/servicegraphconnector/connector.go#L605-L618

It will lead the component to get the wrong label set when it tries to collect metrics:
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/0a12edebb5aad2657c7401c50136d54acc89e573/connector/servicegraphconnector/connector.go#L505

I use a unit test to demonstrate it: [test failed label not work](https://github.com/Frapschen/opentelemetry-collector-contrib/blob/demonstrate_servicegraph_bug/connector/servicegraphconnector/connector_test.go#L96)

In the unit test, I simulate two services' traces data: foo and bar. foo called the bar three times, two successful, and one failed. 
I expect those trace simple will generate graph metrics:
```
traces_service_graph_request_total{client="foo", server="bar", connection_type="", failed="false"} 2
traces_service_graph_request_total{client="foo", server="bar", connection_type="", failed="true"} 1
...
```

however the component result in: [error metrics](https://github.com/Frapschen/opentelemetry-collector-contrib/blob/demonstrate_servicegraph_bug/connector/servicegraphconnector/testdata/bug-metrics.yaml):

<details> <summary>error metrics content</summary>

```yaml
resourceMetrics:
  - resource: {}
    scopeMetrics:
      - metrics:
          - name: traces_service_graph_request_total
            sum:
              aggregationTemporality: 2
              dataPoints:
                - asInt: "3"
                  attributes:
                    - key: client
                      value:
                        stringValue: foo
                    - key: connection_type
                      value:
                        stringValue: ""
                    - key: failed
                      value:
                        boolValue: false
                    - key: server
                      value:
                        stringValue: bar
                  startTimeUnixNano: "1000000"
                  timeUnixNano: "2000000"
              isMonotonic: true
          - name: traces_service_graph_request_failed_total
            sum:
              aggregationTemporality: 2
              dataPoints:
                - asInt: "1"
                  attributes:
                    - key: client
                      value:
                        stringValue: foo
                    - key: connection_type
                      value:
                        stringValue: ""
                    - key: failed
                      value:
                        boolValue: false
                    - key: server
                      value:
                        stringValue: bar
                  startTimeUnixNano: "1000000"
                  timeUnixNano: "2000000"
              isMonotonic: true
```

</details>

In detail, The key problem is that the `metricKey` misses the `failed` label and generates a key that will refer to different values in some cases. 

I can demonstrate it:

firstly, assume this is the first span to go through the connector,  an edge finish with this values(without error): `e.ClientService=foo, e.ServerService=bar,e.ConnectionType=, e.Failed=false`

its metricKey will be `foobar`, then the key refers to its dimensions(stored in a keyToMetric map): `{client:foo, server:bar, connection_type: , failed: false}`
 
currently, the `reqTotal` will be `{"foobar": 1}`, after collect metrics, result metrics will be: 
```
traces_service_graph_request_total{client:foo, server:bar, connection_type: , failed: false} 1
...
```

Then, the second edge finish with this values(contain error): `e.ClientService=foo, e.ServerService=bar,e.ConnectionType=, e.Failed=true`.

This edge also generates the `foobar` and its dimensions will be `{client:foo, server:bar, connection_type: , failed: true}`, after this step, the `foobar`'s value in keyToMetric is overwritten, the bug occurs.  Currently, the `reqTotal` will be `{"foobar": 2}`, and the `reqFailedTotal` will be `{"foobar": 1}`. after collecting, metrics will be:
```
traces_service_graph_request_total{client:foo, server:bar, connection_type: , failed: true} 2
traces_service_graph_request_failed_total{client:foo, server:bar, connection_type: , failed: true} 1
...
```

in the metrics backend, you will see:
```
traces_service_graph_request_total{client:foo, server:bar, connection_type: , failed: false} 1
traces_service_graph_request_total{client:foo, server:bar, connection_type: , failed: true} 2
traces_service_graph_request_failed_total{client:foo, server:bar, connection_type: , failed: true} 1
```

### Collector version

main

## Environment
OS: (e.g., "Ubuntu 20.04")
Compiler(if manually compiled): (e.g., "go 14.2")

### OpenTelemetry Collector configuration

_No response_

### Log output

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Frapschen**
**Description:** <Describe what has changed.>
resolve: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32018

**Link to tracking Issue:**  fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32018

**Testing:** <Describe what testing was performed and which tests were added.>

add a unit test to verify this bug fixed

**Documentation:** <Describe the documentation added

### CODE REVIEW

File: connector/servicegraphconnector/connector.go [hunk 1]
```diff
@@ -395,15 +396,16 @@ func (p *serviceGraphConnector) onExpire(e *store.Edge) {
...
 	dimensions := buildDimensions(e)
 
 	p.seriesMutex.Lock()
 	defer p.seriesMutex.Unlock()
 	p.updateSeries(metricKey, dimensions)
-	p.updateCountMetrics(metricKey)
 	if e.Failed {
 		p.updateErrorMetrics(metricKey)
+	} else {
+		p.updateCountMetrics(metricKey)
```
Comment (mapno) [inline connector/servicegraphconnector/connector.go:408]: This is major change to how the metrics work. `traces_service_graph_request_total` should count all occurrences, not only successful ones.

**Review by mapno — CHANGES_REQUESTED** [review]
Thanks for the investigation (#32018) and for the fix.

I disagree with the proposed solution. I believe the root cause for this is that the metric has a label `failed` in the first place. Failure should either be indicated as a label with a single metric (ie. `traces_service_graph_request_total`), or with a different metric and no label. How it's currently set up is a mix of both, which leaves it a in strange spot.
The way I see forward is doing both approaches.

File: connector/servicegraphconnector/connector.go (hunk 1 above)
Comment (Frapschen) [reply connector/servicegraphconnector/connector.go]: @mapno  Thank you for your careful review, new commit has pushed, please review.

**Review by mapno — APPROVED** [review]
LGTM, thanks!

File: connector/servicegraphconnector/connector.go [hunk 2]
```diff
@@ -610,9 +611,9 @@ func (p *serviceGraphConnector) collectServerLatencyMetrics(ilm pmetric.ScopeMet
 	return nil
 }
 
-func (p *serviceGraphConnector) buildMetricKey(clientName, serverName, connectionType string, edgeDimensions map[string]string) string {
+func (p *serviceGraphConnector) buildMetricKey(clientName, serverName, connectionType string, failed string, edgeDimensions map[string]string) string {
```
Comment (t00mas) [inline connector/servicegraphconnector/connector.go:614]:
```suggestion
func (p *serviceGraphConnector) buildMetricKey(clientName, serverName, connectionType,  failed string, edgeDimensions map[string]string) string {
```

File: connector/servicegraphconnector/connector_test.go [hunk 3]
```diff
@@ -27,6 +27,9 @@ import (
 	"go.opentelemetry.io/otel/sdk/metric/metricdata"
 	"go.opentelemetry.io/otel/sdk/metric/metricdata/metricdatatest"
 	"go.uber.org/zap/zaptest"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/golden"
```
Comment (t00mas) [inline connector/servicegraphconnector/connector_test.go:31]: I like that you introduced `golden` to these tests 💯

**Review by t00mas — APPROVED** [review]
Just left a minor suggestion, if you don't mind

**Review by jpkrohling — APPROVED** [review]

**Review by jpkrohling — APPROVED** [review]

**Comment by andrzej-stencel** [comment]
The connector's unit tests are failing? https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/9347048326/job/25723257817?pr=32019

> Error: ./connector_test.go:101:11: assignment mismatch: 1 variable but newConnector returns 2 values

**Pull request merged**